### PR TITLE
feat(efb): flyPadOS 3 🚀

### LINKED ISSUE #3297: [REQUEST] Show PAX and Cargo Loading on the EFB
**Is your feature request related to a problem? Please describe.**
no

**Describe the feature you'd like to see implemented**

I think when you start up the plane you can hit "load cargo" or "load pax" and watch the pax load in real time as the weight increases as they board. this would add a ton of realism 

**References**

**Additional context**

other aircraft have this feature in xplane 11

Discord username (if different from GitHub): Zyonix007

### LINKED ISSUE #3790: [REQUEST] Open Documents on EFB
It's hard to open another window to open a chart during a flight.

I want Flypad to be able to display PDF, txt, jpg, and png files.

Sorry, I don't have a picture of the actual device.

Discord username: rize_chocolate

### LINKED ISSUE #4443: Checklists on EFB
**Is your feature request related to a problem? Please describe.**

Can we get checklist on EFB? I dont think its that hard to implement since its already on the toolbar. It would be great. 
**Describe the feature you'd like to see implemented**

**References**

**Additional context**

Discord username (if different from GitHub):

**BY REMOVING THIS LINE, I ATTEST THAT I HAVE FOLLOWED THE ISSUE TEMPLATE AND PROVIDED A TITLE, as well as searched for other possible duplicate issues. I acknowledge if this line is not removed and/or the issue template is not respected, my issue will be closed without warning**

### LINKED ISSUE #4601: EFB/FlyPad light/dark mode.
**Is your feature request related to a problem? Please describe.**
No, just a "nice to have" feature request.

**Describe the feature you'd like to see implemented**
I think it would be cool to have an option for the EFB FlyPad to have a light and dark mode themes, kind of like how an iPad has a light and dark mode.

**References**
N/A

**Additional context**
Again, similar to how an iPad has the option to use either a light or dark mode, which can help viewability in some lighting conditions.

Discord username (if different from GitHub): N/A

### LINKED ISSUE #4801: Forward pushback button EFB
### Aircraft Version

Experimental

### Description

The EFB has three directional buttons for controlling the direction of pushback: straight backwards, left and right. However, there is no option for a forward direction, which the [PushbackHelper](https://github.com/metindikbas/msfs-pushback-helper-app) tool does have. It can be handy if you for example overshoot the taxiway centerline and want to correct the alignment of the aircraft easily.

### References (optional)

_No response_

### Additional info (optional)

_No response_

### Discord Username (optional)

Erik Vroon#0182

### LINKED ISSUE #4840: Settable Cockpit States
### Aircraft Version

Development

### Description

Store default cockpit setups (i.e. cold and dark, turn around etc) and have the ability to load these states in sim or set a default start up.

### References (optional)

_No response_

### Additional info (optional)

Aerosoft's A320 implementation
Capture PNG 6f92b56a74c2b42818d343732c20fb8c

### Discord Username (optional)

_No response_

### LINKED ISSUE #5073: Version/Changelog in EFB
### Aircraft Version

Development

### Description

 **About** section in the EFB which has the current version (build if possible, but version is fine), which would show a change-log pop-up with the latest changes from the last version (i.e. 0.6.0 -> 0.7.0). 
 
Users will be informed in this way of the latest changes from the last stable version of A32NX.  Not every change is obvious and some may be buried in options menus, this is a good way to inform casual users who may not be keeping up with the discord/websites/github.

After a new version release, this changelog pop-up can be dismissed and will not show up again, until the version # is updated.

### References (optional)

_No response_

### Additional info (optional)

_No response_

### Discord Username (optional)

2Cas#1022

### LINKED ISSUE #5307: Navigraph charts page in EFB not cached
### Aircraft Version

Development

### Build info

```json
{
  "built": "2021-07-12T22:06:06+00:00",
  "ref": "master",
  "sha": "cf88fc3c16e212978f26186cebfb6dc7deef7af1",
  "actor": "hiaaryan",
  "event_name": "manual"
}
```

### Describe the bug

The Navigraph charts are not cached. So every time you leave the charts page and re-engage it, one has to dial in the ICAO code for the desired airport again. 

### Expected behavior

Cached charts

### Steps to reproduce

Start EFB, go to charts page, enter ICAO code, leave charts page

### References (optional)

_No response_

### Additional info (optional)

_No response_

### Discord Username (optional)

GrandZero#5475

### LINKED ISSUE #5452: Keep current chart open when changing flypad tab
### Aircraft Version

Development

### Description

Hey guys!
It would be cool to keep the latest  navigraph chart open when selecting another tab in flypad. At the current situation when selecting thr e.g. ground tab and then the charts tab again, you completely have to type in the airport and search your chart again.

Maybe its a misusing on my side, but having that would be great and would enhance the atmosphere!

Greetings!

### References (optional)

_No response_

### Additional info (optional)

_No response_

### Discord Username (optional)

_No response_

### LINKED ISSUE #5522: Pushback steering via yaw input
### Aircraft Version

Development

### Description

It would be nice if we can steer the pushback left and right via the yaw input.

### References (optional)

_No response_

### Additional info (optional)

_No response_

### Discord Username (optional)

KeepGamingLP#4346

### LINKED ISSUE #5604: get rid of fake efb loading screen
### Aircraft Version

Development

### Description

We should get rid of the fake efb loading screen. #getRidOfFakeLoadingScreen 

### References (optional)

_No response_

### Additional info (optional)

_No response_

### Discord Username (optional)

_No response_

### LINKED ISSUE #5615: FlyPad battery life
### Aircraft Version

Stable

### Description

If you look up at the top right corner of the Flypad, you see the battery meter and how it is always charging. I think it should be that while AC power isn't being supplied to the plane (Both APU and exterior power are off) the flypad doesn't charge and the battery starts to drain, like how the flight instruments aren't powered.

### References (optional)

_No response_

### Additional info (optional)

Similar to how the flight instruments are not powered when neither the APU or EXT PWR is avail, the same should be for the FlyPad's charging supply.

### Discord Username (optional)

_No response_

### LINKED ISSUE #5632: EFB Routing and State Overhaul
### Aircraft Version

Development

### Description

Currently, there are inconsistencies in Routing and State Management that need to be addressed. 

Standards need to be implemented for all contributors to adhere to in order to reduce the introduction of bespoke patterns for new features/enhancements

### References (optional)

_No response_

### Additional info (optional)

_No response_

### Discord Username (optional)

antiso

### LINKED ISSUE #5734: flyPad: Landing Calculator Unit Toggles
### Aircraft Version

master

### Description

Add toggles to some/all of the units in the landing calculator. For example, QNH in mb to in Hg. Big one is LDA to feet, since all the US charts reference runway lengths in feet.

Ideally this would also toggle the output into feet as well.

The kg/LB toggle could probably use the setting already present in the flyPad settings.

### References (optional)

image

### Additional info (optional)

_No response_

### Discord Username (optional)

Kennedy#6132

### LINKED ISSUE #6160: [EFB Request] Add a way for users to manually add a certain value for EFB Audio Settings
### Aircraft Version

Experimental

### Description

image

On the EFB Audio settings, I found that it could be quite annoying trying to get a certain value by using the slider. I think it would be more user-friendly to add a way that users can input a value from -100 to +100.  Maybe if the numbers on the left of the sliders are clicked, then you can input a certain value. 

### References (optional)

_No response_

### Additional info (optional)

_No response_

### Discord Username (optional)

GhostEagle68#0030

### LINKED ISSUE #6230: Option to configure PFD's/Screen Brightness using config files or EFB
### Aircraft Version

Development

### Description

I wish we could configure PFD's or Screen Brightness using configuration file or the EFB to save the effort and process of using the brightness knobs.
Sometimes, launching the flight in bad weather or full day time , will require large brightness adjustments.
I find my self constantly adjusting the brightness on each flight and will love to set those settings in s pre defined state (such as configuration files) or use the EFB for fast brightness settings.

Thanks !

### References (optional)

_No response_

### Additional info (optional)

_No response_

### Discord Username (optional)

Eran.S WW5446#2617

### LINKED ISSUE #6263: Add option to import fuel from simbrief to EFB
### Aircraft Version

Development

### Description

With the new w&b and fuel management one has to manually select the desired fuel on the dispatch /fuel page in the EFB.
Could you add an "import fuel from simbrief" button to that page?

### References (optional)

_No response_

### Additional info (optional)

_No response_

### Discord Username (optional)

foxmike1080

### LINKED ISSUE #6302: Distinct buttons for DEP, DEST, ALT
### Aircraft Version

Development

### Description

On the FlyPad Navigraph Chart page - Add buttons for direct selection of the DEP, DEST, ALT airports so no keyboard input is needed.

In particular in VR that would be an enourmous help.

### References (optional)

_No response_

### Additional info (optional)

_No response_

### Discord Username (optional)

sso#6947

### LINKED ISSUE #6343: Add autoland (yes or no) field to EFB landing distance calculator
### Aircraft Version

Development

### Description

The landing distance calculator provides in-flight landing distance data to assist pilots in determining if they can safely land on a particular runway. It includes inputs for all of the variables that affect the in-flight landing distance computation.  However, it is missing one item, and that is whether or not an autoland is being conducted.

There should be an additional input box for whether or not autoland is being used. It should have 2 choices, (1) yes or (2) no.

### References (optional)

See this page from the FCOM on which the landing distance calculator is based. I have marked the missing autoland landing distance correction with a blue arrow.
Screenshot (1386)_LI

### Additional info (optional)

I should have noticed this when reviewing the PR introducing the landing distance calculator, but I missed it.

### Discord Username (optional)

donbikes#4084

### LINKED ISSUE #6478: Incorrect calculation of VLS in EFB Landing Distance Calculator
### Aircraft Version

Development

### Build info

```json
{
  "built": "2022-01-03T21:08:06+00:00",
  "ref": "master",
  "sha": "5bd7bc97482f6cdc326bbde9efe93d0ed857cc95",
  "actor": "aguther",
  "event_name": "manual"
}
```

### Describe the bug

The EFB Landing Distance Calculator includes a correction to the landing distance for speed when the input approach speed is above the VLS for the selected landing flap setting. VLS speeds are defined in the calculator as a function of weight for each of the 2 possible landing flap settings. VLS speeds are provided every 5 tonnes for weights between 40 tonnes and 80 tonnes.

The issue is that the calculator does not interpolate between the 5 tonne weight increments for VLS. It uses the VLS for a given weight for all weights between it and the next weight increment.  This is best illustrated by an example. For FULL landing flaps, VLS is defined in the calculator as 120 KCAS at 55T and 125 KCAS at 60T. By interpolation, VLS at 56T should be 121 KCAS, at 57T, VLS should be 122 KCAS, at 58T, VLS should be 123 KCAS, and at 59T, VLS should be 124 KCAS. However, the landing calculator calculates VLS at 55T through 69T as 120 KCAS. 

This can be observed without having an output of the calculated VLS. Just observe that the calculated landing distance does not increase for Flaps FULL at 56T when the approach speed is changed from 121 KCAS (VLS) to 122 KCAS. The landing distance does not increase until the approach speed is changed to 126 KCAS, indicating that the calculator "thinks" VLS for 56T is 125 KCAS. (Remember:  the calculator adds distance whenever the approach speed is greater than VLS.)

### Expected behavior

The calculator should interpolate between the weights to obtain the correct VLS between table values. 

For the example above, with FULL flaps and a landing weight of 56T, the landing distance should increase when the approach speed is increased above 121 KCAS.

### Steps to reproduce

See above

### References (optional)

_No response_

### Additional info (optional)

_No response_

### Discord Username (optional)

donbikes#4084

### LINKED ISSUE #6602: [EFB] Keep the ICAO METAR on front page (when inserted by hand)
### Aircraft Version

Development

### Description

For Dev and Exp version: Right now, if you **_don't_** use SimBrief import on the EFB and type your departure and destination in by hand (METAR > front page of EFB), the EFB forgets the ICAO codes / METARs when switching to another page or turning the EFB off for a while during flight/session.

I would suggest to make the EFB remember this information during a session - when switching pages but also when turning the EFB off during the flight / the same session. The codes could be cleared automatically once the plane or MSFS is shut down.  

### References (optional)

Nope

### Additional info (optional)

_No response_

### Discord Username (optional)

RogePete

### LINKED ISSUE #6618: EFB: Show Sim time instead of Computer time
### Aircraft Version

Development

### Description

Hi guys,
maybe just a small detail, but would be more realistic, EFB time would show sim-time and not the computer time.

### References (optional)

_No response_

### Additional info (optional)

_No response_

### Discord Username (optional)

aighes

### LINKED ISSUE #6627: EFB Charts - save viewed page if switch to other functions and back
### Aircraft Version

Experimental

### Description

It would be awesome if the "charts page" on the EFB could save the state, if you switch e.g. to fuel  or OFP - page and back.

now you must select the airport and respective page everytime of you go to CHART-page.

### References (optional)

_No response_

### Additional info (optional)

_No response_

### Discord Username (optional)

_No response_

### LINKED ISSUE #7246: Pushback via EFB going in wrong direction upon first click
### Aircraft Version

Development

### Build info

```json
{
    "built": "2022-05-25T01:03:21+00:00",
    "ref": "refs/heads/master",
    "sha": "bba5b4acc558c277c021082e749a20001aecee4b",
    "actor": "2hwk",
    "event_name": "manual",
    "pretty_release_name": "master:bba5b4ac"
}
```

### Describe the bug

The tug goes into the opposite direction when clicking a certain direction on the EFB. Upon the second click it will correct itself. Further clicks seem to mess it up as well (might be alternating or something, not sure).

### Expected behavior

Tug should turn in the expected direction upon the first click.

### Steps to reproduce

1. Start pushback
2. Push left or right arrow once, observe wrong turning direction.
3. Push same error a second time, observe correction into the right direction.

### References (optional)

https://www.twitch.tv/videos/1310637697?t=00h07m40s

### Additional info (optional)

_No response_

### Discord Username (optional)

_No response_

### PR DESCRIPTION
**Pull request by ErickSharp**
Fixes #3297
Fixes #3790
Fixes #4443
Fixes #4601
Fixes #5073
Fixes #5307 
Fixes #5452
Fixes #5604
Fixes #5615
Fixes #5632
Fixes #5734
Fixes #6160
Fixes #6263
Fixes #6343
Fixes #6478
Fixes #6602
Fixes #6618 
Fixes #6627
Fixes #6230
Fixes #4840
Fixes #4801
Fixes #6302
Fixes #5522
Fixes #7246

## Summary of Changes

flypadOS 3 aims to unify the design language, code structure, and conventions present within the EFB. This is all in the hope of giving users a better experience, and giving developers a more consistent and easier codebase to work with. Landmark features include redone state management, updated page designs, theming, a battery charge/discharge implementation, and new functional pages like the checklist page.

- [x] Redone State Management
- [x] Notifications
- [x] Modals
- [x] More Power States
- [x] Redone Settings Page
- [x] Chart Bookmarking
- [x] Landing Calculator Redesign
- [x] Redone Dispatch Pages
- [x] Redone Flight Widget
- [x] Refreshed Weather Widget (Currently Being Worked On)
- [x] Theming
- [x] Checklist Page
- [x] Localization - (ongoing)
- [x] Battery Life
- [x] On-Screen Keyboard (WIP)
- [x] Presets for Interior Lighting and Aircraft States
- [x] Local File Charts
- [x] Pushback 
  - [x] Pushback control for direction and speed with smooth movements
  - [x] Pushback UI with direction, speed, pause, parking brake, call/release tug, enable/disable system
  - [x] Compatibility Mode with other pushback add-on (disable system)
  - [x] Map (draggable, moving map centered on aircraft, moving aircraft)
  - [x] Bug Fixing:
    - [x] Lukla Test - nose wheel in the air at edge
  - [x] Testing, debugging and tuning 
- [x] Bug Fixing
  - [x] [TOD Calculator Issue](https://discord.com/channels/738864299392630914/957208413887139860/972643025891246120) 
  - [x] Landing Performance Calculator - Runway distance marking - text vanishes
  - [x] Autobrightness needs improvement (will be PR)
- [x] Document pushback LVAR
- [x] Clean up / remove unfinished languages
- [x] Finishing Touches

**Remove local files and health check do to simbridge not yet merged - code is only commented out.**

## Testing instructions
General final testing of the whole flyPad before merging into Development.
Docs preview:https://docs-git-fork-frankkopp-flypados3-docs-flybywire.vercel.app/fbw-a32nx/feature-guides/flyPados3/

- Focus on testing on each page and subpage - see documentation if it is unclear how to use a feature. 
- Look at performance impact using the dev mode fps counter. Check CoherentGTUIThread 90% value when using flyPad or when it is off (the difference is of interest). Also check impact on fps for you. More specifically check the impact on the CoherentGTUIThread 90% value between pushback system enabled and disabled.
- Also check if documentation is correct 
- If you can check translation into your own language
- Try to break it 🙂

Feedback in the dedicated Discord channel and/or report here as usual. 

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

**Comment by wpine215** [comment]
When you get the chance, can you add a checkbox list indicating what's finished and what's left to be done? Would be helpful to anyone who wants to help out with the remaining items so this can be merged soon,

**Comment by ErickSharp** [comment]
> When you get the chance, can you add a checkbox list indicating what's finished and what's left to be done? Would be helpful to anyone who wants to help out with the remaining items so this can be merged soon,

Sure thing, whenever I have some time later today I'll get a better description up with some screenshots.

**Comment by ErickSharp** [comment]
Just a disclaimer: I have a couple or so great contributors willing to spend some of their time to work alongside me. If they are still working and the majority of things on the list are complete, I may be able to squeeze one or two more things in here.

**Comment by ErickSharp** [comment]
Failures page redesign is coming soon but won't be within this pull request as I need to edit files that would likely be better off in a separate pull request for review.

**Comment by donstim** [comment]
What does a redesign mean specifically, like the landing calculator redesign? Is it just to clean up the code for future improvements to the page, or is it to also fix issues and improve the existing calculator?

For example, there are existing issues #6478, #6343, and #5734. (Oh, I see that #5734 is listed in the description as being fixed.)

Others for which no issues have been opened (I can open them if you'd like) include:
- Rather than just having an input for the ICAO code for the airport also have a dropdown for the runways at that airport. Upon selection of the runway, autofill the runway heading, runway altitude, Rwy slope, and Rwy LDA.
- Autofill the Approach Speed input with the approach speed from the MCDU PERF APPR page (but also allow it to be overwritten)
- Maybe allow the Rwy Condition to be obtained from the ATIS if it is provided

**Comment by donstim** [comment]
Will the boarding page include all of the improvements that have been discussed in the various Discord channels, or will this be done in a later PR?  If in this PR, can the specific items be listed here, so that we can verify that it is a complete list?

**Comment by ErickSharp** [comment]
> Will the boarding page include all of the improvements that have been discussed in the various Discord channels, or will this be done in a later PR? If in this PR, can the specific items be listed here, so that we can verify that it is a complete list?

Depending on how much time other contributors and I have, we may be able to get further in depth visualizations and add more features but for now I think we will get a port of what is currently present in the MCDU with a new UI.

**Comment by ErickSharp** [comment]
> What does a redesign mean specifically, like the landing calculator redesign? Is it just to clean up the code for future improvements to the page, or is it to also fix issues and improve the existing calculator?
> 
> For example, there are existing issues #6478, #6343, and #5734. (Oh, I see that #5734 is listed in the description as being fixed.)
> 
> Others for which no issues have been opened (I can open them if you'd like) include:
> 
> * Rather than just having an input for the ICAO code for the airport also have a dropdown for the runways at that airport. Upon selection of the runway, autofill the runway heading, runway altitude, Rwy slope, and Rwy LDA.
> * Autofill the Approach Speed input with the approach speed from the MCDU PERF APPR page (but also allow it to be overwritten)
> * Maybe allow the Rwy Condition to be obtained from the ATIS if it is provided

Redesign initially meant just updating the design of the page and cleaning up the code a bit. However, I can fix some of the issues that you mentioned. I want the drop down feature myself but I still need to find a source to get this data in the first place.

**Comment by donstim** [comment]
>> I want the drop down feature myself but I still need to find a source to get this data in the first place.

Navigraph should have it (as well as the LDA), but I don't know anything about where they provide it. I would think it should be available from some database in the base sim as well in order to do flight planning, etc.

**Comment by tracernz** [comment]
> source to get this data in the first place

> runway heading, runway altitude, Rwy slope, and Rwy LDA

Runway heading and elevation is directly in the nav data. Also in there are total runway length, threshold displacement for each end, and threshold elevations for each end, from which you can calculate slope, and an approximation of LDA.

We already do the slope calculation in https://github.com/flybywiresim/a32nx/blob/master/flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/FlightElements/A32NX_Runway.js#L32. If you can't wait this would be the data to use... if you can wait until we integrate msfs-navdata that would be the best as it wouldn't need reworked later.

**Comment by donstim** [comment]
For the LDA and runway slope, see the Discord discussions here: https://discord.com/channels/738864299392630914/805556372904476672/808635978499358740, here: https://discord.com/channels/738864299392630914/805556372904476672/864921275952529418 and here: https://discord.com/channels/738864299392630914/805556372904476672/867464481852096532

**Comment by tracernz** [comment]
> Or the relevant text copied and pasted here [Navigraph runway data.txt](https://github.com/flybywiresim/a32nx/files/7999332/Navigraph.runway.data.txt) :

When I say "if you can wait until we integrate msfs-navdata"... [msfs-navdata](https://github.com/flybywiresim/msfs-navdata) is the abstraction layer we're currently building that allows us to use either msfs native navdata from the sim, or raw navigraph data from them. 😉

**Comment by davidwalschots** [comment]
I'd like to note that there are various other PRs/branches which are based off of `flypadosv3`. Every time you force push your commits it creates extra work for those authors. Consider not rebasing to reduce the burden on your fellow authors.

**Comment by ErickSharp** [comment]
> I'd like to note that there are various other PRs/branches which are based off of `flypadosv3`. Every time you force push your commits it creates extra work for those authors. Consider not rebasing to reduce the burden on your fellow authors.

This is indeed something that I already consider, but Mav and I were ok with doing this most recent rebase. I find that if I don't rebase at least once every week to two weeks then I get crushed with all the incoming changes. I will keep it under consideration though and try to get everyone on the same page!

**Comment by TayDex** [comment]
The Navigraph charts are quite low resolution anyway, this could be fixed/changed?

Navigraph charts panel vs EFB:
lowres

**Comment by Benjozork** [comment]
@ErickSharp

**Comment by ErickSharp** [comment]
> 

The resolution of the image is unfortunately something we are unable to control when we request the image from Navigraph servers.

**Comment by ZigTag** [comment]
> > 
> 
> The resolution of the image is unfortunately something we are unable to control when we request the image from Navigraph servers.

Actually it could be changed, you'd need to use the PDF versions of the Charts, however, I don't know if you'd be able to overlay the live plane

**Comment by ErickSharp** [comment]
> > > 
> > 
> > 
> > The resolution of the image is unfortunately something we are unable to control when we request the image from Navigraph servers.
> 
> Actually it could be changed, you'd need to use the PDF versions of the Charts, however, I don't know if you'd be able to overlay the live plane

We have had issues in the past trying to use a pdf-viewing React library so that's the reason the local API serves pdfs as images. Maybe we could do something similar for the fetched PDFs?

**Comment by frankkopp** [comment]
We really should not squash this branch! 
I understand that this is easier to merge to Exp but then one should do a branch copy, squash and cherry pick the squashed commit. 
We loose all history in this branch with each squash.

**Comment by Maximilian-Reuter** [comment]
As it stands right now the EFB seems to consume power even while off, leading to situations where it is empty after changing the sim time while C/D which is unrealistic and annoying.

Also the entire battery charging/discharging feature right now can lead to situations where you are unable to reset a ELEC failure if the EFB is low on power. Which will only be exacerbated if persistence is applied to it.

As such I would suggest a kind of emergency power mode that at least allows access to the failure page.

For further ref see discord conversation: https://discord.com/channels/738864299392630914/747862215679279214/960821125108084776

File: src/instruments/src/EFB/Performance/Widgets/RunwayVisualizationWidget.tsx [hunk 1]
```diff
@@ -45,11 +45,15 @@ interface RunwayNumberProps {
     heading?: number;
 }
 
-const RunwayNumber = ({ heading }: RunwayNumberProps) => (
-    <div className="mx-auto w-min text-4xl font-bold text-white">
-        {heading !== undefined ? Math.round((heading % 360 ?? 0) / 10).toString().padStart(2, '0') : '??'}
-    </div>
-);
+const RunwayNumber = ({ heading }: RunwayNumberProps) => {
+    const displayedHeading = heading === 360 ? 360 : heading! % 360;
```
Comment (tracernz) [inline src/instruments/src/EFB/Performance/Widgets/RunwayVisualizationWidget.tsx:49]: 0 should also be runway 36.

**Comment by svengcz** [comment]
We need to remove the Hoppie-connect button and the Datalink entry in the realism page.

**Comment by ErickSharp** [comment]
> We need to remove the Hoppie-connect button and the Datalink entry in the realism page.

Can you give some context? Why?

**Comment by svengcz** [comment]
> > We need to remove the Hoppie-connect button and the Datalink entry in the realism page.
> 
> Can you give some context? Why?

It's not needed anymore with the last merge. The system automatically logs in and we have a realistic simulation of the VHF-transmission protocol which cannot be deactivated. I will prepare a commit for it

**Comment by ErickSharp** [comment]
> 

Ok 👍

**Comment by frankkopp** [comment]
FYI Hoppie language key removal and adding is done

**Comment by ErickSharp** [comment]
> FYI Hoppie language key removal and adding is done

Awesome 😋

**Review by frankkopp — APPROVED** [review]

**Comment by ErickSharp** [comment]
WE DID IT!!!

**Pull request merged**